### PR TITLE
Tests and debug

### PR DESCRIPTION
**Pull request by kaikun213**
* Extended CP tests with more toy-examples
* Activation was because wrong function call (L2Representations instead of L2Prediction) in test
* Parameters were only used in "Summing"-version
* Problem: We learn internalDistalConnections to timestep t but should learn to t-1. (To infer an object also based on the previous sensations/locations) E.g. as described in Numenta paper: 
"Thus, if a column
has feedforward support for objects A and B at time t, and
feedforward support for objects B and C at time t+1, the output
layer will converge onto the representation for object B at time
t+1 due to modulatory input from time t."
=> Fix: Use t-1 activation for internalDistal and change predictedObject logic to include support&activation (multiply probabilities currently)
* Problem: the activation function with ignoreNinf=False, ignores that if we have neurons active from two objects, they might not have any weight to each other. Thus their activations would be an array with values e.g. [       -inf  0.05384701  0.05384701  0.05384701  0.03164526  0.05384701]. Where one other neuron was active in input, but did not have a connection. -inf will be the result when summing the values and thus put the activation of the neuron to 0.
However, with ignoreNinf activated we will have them all masked out and exp(0) = 1. Thus we should only mask out if there are actual values. => Fixed by explicitly setting to zero

### CODE REVIEW

File: htmresearch/algorithms/bayesian_column_pooler_base.py [hunk 1]
```diff
@@ -408,10 +418,16 @@ def _activation(weights, input, bias, noise, useBias=True, ignoreNinf=False):
...
-        if not ignoreNinf:
-            activation = activation.sum(axis=1) if np.any(activeMask) else activation.sum(axis=1) + np.NINF
-        else:
-            activation = np.ma.masked_invalid(activation).sum(axis=1) if np.any(activeMask) else activation.sum(axis=1) + np.NINF
+        # if not ignoreNinf:
+        #     activation = activation.sum(axis=1) if np.any(activeMask) else activation.sum(axis=1) + np.NINF
+        # else:
+        #     activation = np.ma.masked_invalid(activation).sum(axis=1) if np.any(activeMask) else activation.sum(axis=1) + np.NINF
+
+        # Set explicitly to zero to avoid summing NINF if only one connection is non-existend for a neuron (because multiple objects are activated and only overlapping would be inferred)
```
Comment (leoTiez) [inline htmresearch/algorithms/bayesian_column_pooler_base.py:426]: That's something what I'd like to change (is already included in the other PR so it's fine for now and nothing what prevents us from merging). I'd prefer using standard matrix multiplication. That requires valid values for all weights. Then we don't end up with special handling for -inf

File: htmresearch/regions/BayesianColumnPoolerRegion.py [hunk 2]
```diff
@@ -472,15 +472,16 @@ def initialize(self):
...
-        params["forgetting"] = self.forgetting
-        params["useSupport"] = self.useSupport
-        params["avoidWeightExplosion"] = self.avoidWeightExplosion
-        params["resetProximalCounter"] = self.resetProximalCounter
-        params["useProximalProbabilities"] = self.useProximalProbabilities
-
       else:
         raise ValueError("Unrecognized implementation %s" % self.implementation)
 
+      # Params are needed in both implementations (base version)
```
Comment (leoTiez) [inline htmresearch/regions/BayesianColumnPoolerRegion.py:478]: Well seen

File: tests/bayesian/experiment_suite.py [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+    learning_rate_l2_p = [0.1, 0.2]
+    learning_rate_l4_p = [0.1, 0.2]
+    activation_threshold_p = [0.3, 0.5, 0.7]
+    min_threshold_p = [0.1, 0.3, 0.5]
+    forgetting_p = [0.01, 0.1, 0.15]
+    use_support_p = [True, False]
+    use_proximal_probabilities_p = [True, False]
+    use_apical_tiebreak_p = [True, False]
+
+    for num_columns in num_columns_p:
```
Comment (leoTiez) [inline tests/bayesian/experiment_suite.py:34]: These convoluted loops are unnecessary. That makes it really hard to read the code. Use instead:
`itertools.product` and then loop only over the single list.
https://docs.python.org/2/library/itertools.html#itertools.product

File: tests/bayesian/experiment_suite.py [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+                                                numCorticalColumns=num_columns,
+                                                maxSegmentsPerCell=maxNumSegments,
+                                                seed=1,
+                                                L2Overrides=L2Overrides,
+                                                L4Overrides=L4Overrides,
+                                            )
+
+                                            with warnings.catch_warnings():
+                                                if (suppressWarnings):
+                                                    warnings.simplefilter("ignore")
```
Comment (leoTiez) [inline tests/bayesian/experiment_suite.py:91]: Does this catch any error thrown and continues with the next parameter setting, or does it only suppress warnings whereas errors still stop the script to execute? It is important to continue since some parameter settings just might not make any sense in their configuration and hence bring the execution to fail. If we don't catch this case we won't be able to go through all of the settings and to run all our experiments.

File: tests/bayesian/test_baysian_cp_full.py [hunk 5]
```diff
@@ -14,10 +14,14 @@ def setUp(self):
...
             "noise": 1e-8,
-            "cellCount": 256,  # new: 256 # original: 4096
+            "cellCount": 16, # 256,  # new: 256 # original: 4096
             "inputWidth": 8192,  # new: 8192 # original: 16384 (?)
             "sdrSize": 5,
-            "activationThreshold": 0.01,
+            "activationThreshold": 0.5,
+            "forgetting": 0.01,
+            "useSupport": True,
+            # "useProximalProbabilities": True,
```
Comment (leoTiez) [inline tests/bayesian/test_baysian_cp_full.py:23]: Is there a reason why this is commented out an not set to a particular value?

File: tests/bayesian/experiment_suite.py [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+
+  # 1.) Inference: Get active predictions for each sensation/location input
+  l2ActiveCells = []
+  for sensation in sensationSteps:
+    exp.infer([sensation], objectName=objectId, reset=False)
+    rep = exp.getL2Prediction()
+    l2ActiveCells.append(rep)
+    activeCellNum = 0
+    for c in range(numColumns):
+        activeCellNum += len(rep[c])
```
Comment (leoTiez) [inline tests/bayesian/experiment_suite.py:184]: Why do we do that? Where do we use `activeCellNum` again?

**Review by leoTiez — CHANGES_REQUESTED** [review]

File: tests/bayesian/test_baysian_cp_full.py (hunk 5 above)
Comment (kaikun213) [reply tests/bayesian/test_baysian_cp_full.py:23]: That's the used test configuration. Tested with both configs, could be set to False alternatively

File: tests/bayesian/experiment_suite.py (hunk 6 above)
Comment (kaikun213) [reply tests/bayesian/experiment_suite.py:184]: At the moment not, but left it in case we use multi-columns, we could check for total activity

File: tests/bayesian/experiment_suite.py (hunk 4 above)
Comment (kaikun213) [reply tests/bayesian/experiment_suite.py:91]: Only suppresses warning. Yes for an error prone run, your setup is better or an additional catch phrase would be needed.
However, this works for simple testing.

File: tests/bayesian/experiment_suite.py (hunk 3 above)
Comment (kaikun213) [reply tests/bayesian/experiment_suite.py:34]: True that's better. :+1:

File: htmresearch/algorithms/bayesian_column_pooler_base.py (hunk 1 above)
Comment (kaikun213) [reply htmresearch/algorithms/bayesian_column_pooler_base.py:426]: Have to check for equivalence. I would prefer keeping it explicit like this at the moment. 
Avoids bugs and makes it very clear which weights are used.

Also because I am not sure yet about the normalization of input values in the PR.

**Comment by leoTiez** [comment]
Close due to deprecation
